api/messages.proto

```diff
@@ -204,6 +204,14 @@ message ApplicationUplink {
   // Warnings generated by the message processor while decoding the frm_payload.
   repeated string decoded_payload_warnings = 12;
 
+  // The normalized frame payload of the uplink message.
+  // This field is set by the message processor that is configured for the end device (see formatters) or application (see default_formatters).
+  // If the message processor is custom script and the decoded payload output is a valid normalized payload structure,
+  // this field is set to the decoded payload.
```

Comment:
So if the `decodeUplink` function returns a result that it's already normalized, the `normalizeUplink` function should not run at all even if it exits? And both fields (`decoded_payload` and `normalized_payload`) will be equal? Where is this logic implemented?

Reply:
> So if the `decodeUplink` function returns a result that it's already normalized, the `normalizeUplink` function should not run at all even if it exits?

If it exists, it will run. The output is either the output from the normalizer, or from the decoder if that output is already normalized. I'll see if I can clarify that in this comment.

> And both fields (`decoded_payload` and `normalized_payload`) will be equal?

Yes

> Where is this logic implemented?

In the Application Server.